controller/volunteer_controller.py

```diff
@@ -27,25 +35,68 @@ def view_volunteer_profile(volunteer: Volunteer) -> str:
     pass
 
 
-def edit_volunteer_profile(volunteer: Volunteer, firstname: str, lastname: str,
-                           phone: str) -> Volunteer:
-    """
-    Yunsy, Yingbo
-    A function used by admin and volunteer
-    Should include re-assign volunteer to a camp
-    """
-    pass
-
-
-def change_volunteer_camp(volunteer: Volunteer, camp: Camp, is_admin: bool) -> Volunteer:
-    """
-    The admin role
-    :param is_admin: if this method is called by an admin
-    :param volunteer:
-    :param camp: the new camp
-    :return:
-    """
-    pass
+def edit_firstname(volunteer: Volunteer, firstname: str) -> Volunteer:
+    try:
+        Volunteer.check_volunteer_firstname(firstname)
+        volunteer.firstname = firstname
+        volunteer.save()
+        return volunteer
+    except Volunteer.InvalidFirstnameException as e:
+        raise ControllerError(str(e))
+
+
+def edit_lastname(volunteer: Volunteer, lastname: str) -> Volunteer:
+    try:
+        Volunteer.check_volunteer_lastname(lastname)
+        volunteer.lastname = lastname
+        volunteer.save()
+        return volunteer
+    except Volunteer.InvalidLastnameException as e:
+        raise ControllerError(str(e))
+
+
+def edit_phone(volunteer: Volunteer, phone: str) -> Volunteer:
+    try:
+        Volunteer.check_volunteer_phone(phone)
+        volunteer.phone = phone
+        volunteer.save()
+        return volunteer
+    except Volunteer.InvalidPhoneException as e:
+        raise ControllerError(str(e))
+
+
+def edit_camp(volunteer: Volunteer, camp: Camp, is_admin: bool) -> Volunteer:
+    old_camp = volunteer.camp
+    if is_admin:
+        old_camp.volunteers.remove(volunteer)
+        old_camp.save()
+        camp.volunteers.add(volunteer)
+        camp.save()
+        return volunteer
+    else:
+        if old_camp.plan == camp.plan:
+            old_camp.volunteers.remove(volunteer)
+            old_camp.save()
+            camp.volunteers.add(volunteer)
+            camp.save()
+            return volunteer
+        else:
+            raise ControllerError(f"Invalid camp: {camp}. You can only select a camp under the same plan.")
+
+
+def edit_availability(volunteer: Volunteer, availability: str) -> Volunteer:
+    list_true = ["True", "true", "T", "1"]
+    list_false = ["False", "false", "F", "0"]
```

Comment:
I think str conversion would better be handled in the interface

Reply:
already relocated the str conversion to edit_volunteer_cli.py